Detectors/TPC/simulation/src/DigitContainer.cxx

```diff
@@ -13,37 +13,50 @@
 /// \author Andi Mathis, TU München, andreas.mathis@ph.tum.de
 
 #include "TPCSimulation/DigitContainer.h"
-#include "TPCBase/Mapper.h"
 #include <iostream>
 
 using namespace o2::TPC;
 
-void DigitContainer::addDigit(size_t hitID, int cru, int timeBin, int row, int pad, float charge)
+void DigitContainer::addDigit(size_t eventID, size_t hitID, const CRU &cru, TimeBin timeBin, GlobalPadNumber globalPad, float charge)
 {
-  /// Check whether the container at this spot already contains an entry
-  DigitCRU *result = mCRU[cru].get();
-  if(result != nullptr){
-    mCRU[cru]->setDigit(hitID, timeBin, row, pad, charge);
+  const Sector sector = cru.sector();
+  if(mSectorID == sector) {
+    mSector.setDigit(eventID, hitID, cru, timeBin, globalPad, charge);
   }
-  else{
-    const Mapper& mapper = Mapper::instance();
-    mCRU[cru] = std::make_unique<DigitCRU>(cru, mCommonModeContainer);
-    mCRU[cru]->setDigit(hitID, timeBin, row, pad, charge);
+
+  // stupid fix for now to fill the container
+  if (cru.number() == 11){
+    mOutOfSector.emplace_back(DigitOutOfSector(eventID, hitID, sector, timeBin, globalPad, charge));
   }
-  /// Take care of the common mode
-  mCommonModeContainer.addDigit(cru, timeBin, charge);
 }
 
 
-void DigitContainer::fillOutputContainer(std::vector<o2::TPC::Digit> *output, o2::dataformats::MCTruthContainer<o2::MCCompLabel> &mcTruth,
-                                         std::vector<o2::TPC::DigitMCMetaData> *debug, int eventTime, bool isContinuous)
+void DigitContainer::fillOutputContainer(std::vector<Digit> *output, dataformats::MCTruthContainer<MCCompLabel> &mcTruth,
+                                         std::vector<DigitMCMetaData> *debug, TimeBin eventTime, bool isContinuous)
+{
+  /// \todo Use CRU here instead of the sector!
+  mSector.fillOutputContainer(output, mcTruth, debug, Sector(mSectorID), eventTime, isContinuous);
+}
+
+void DigitContainer::fillOutOfSectorContainer(std::vector<Digit> *output, dataformats::MCTruthContainer<MCCompLabel> &mcTruth)
 {
-  for(auto &aCRU : mCRU) {
-    if(aCRU == nullptr) continue;
-    aCRU->fillOutputContainer(output, mcTruth, debug, aCRU->getCRUID(), eventTime, isContinuous);
-    if(!isContinuous) {
-      aCRU->reset();
+  const static Mapper &mapper = Mapper::instance();
+
+  /// sort according to sector, time bin and global pad
+  std::sort(mOutOfSector.begin(), mOutOfSector.end());
+
+  /// merge digits on the same sector, time bin and global pad
+  for(auto it = mOutOfSector.begin(); it < mOutOfSector.end(); ++it) {
+    const size_t digiPos = output->size();
+    if(*it == *(it-1)) {
```

Comment:
My problem here is that I'm not sure where `it-1` will point to in case `it` is the first element. If you think of it as a pointer:
`int val[10]; auto it = val;` then `it-1` will point to an invalid position. I'm not sure if this can happen here as well and, under some circumstances, crash.